Sphinx

BalzaniEdoardo opened the pull request:
Switch to sphinx pydata theme.

Adding some notes here on the docstrings parsers:

- numpy uses the extension `numpydoc` for docstrings parsing; we are using another sphinx extension called `napoleon`. `numpydoc` works best in parsing, way less unexpected rendering (for example, multi-line exceptions renders fine), but it doesn't support type hinting (type hinting are dropped). This is because numpy doesn't support type hinting too (see here for a discussion: https://github.com/numpy/numpydoc/issues/196). In the future this may change and I will strongly recommend to switch to `numpydoc`. Napoleon has a clear bug in `sphinx/ext/napoleon/docstring.py`:
```python
    def _parse_raises_section(self, section: str) -> list[str]:
        fields = self._consume_fields(parse_type=False, prefer_type=True)
        lines: list[str] = []
        for _name, _type, _desc in fields:
            m = self._name_rgx.match(_type)
            if m and m.group('name'):
                _type = m.group('name')
            elif _xref_regex.match(_type):
                pos = _type.find('`')
                _type = _type[pos + 1 : -1]
            _type = ' ' + _type if _type else ''
            _desc = self._strip_empty(_desc)
            _descs = ' ' + '\n    '.join(_desc) if any(_desc) else ''  # here they are creating a string that doesn't render as HTML
            lines.append(f':raises{_type}:{_descs}')
        if lines:
            lines.append('')
        return lines
```
 but I don't know what to do about it. For now, I split the exception into multiple (i.e. a ValueError with few cases, is split into multiple errors of one case).

billbrod commented on src/nemos/observation_models.py:
```diff
@@ -23,8 +23,9 @@ class Observations(Base, abc.ABC):
-    of the abstract methods: _negative_log_likelihood, emission_probability, and
-    residual_deviance.
+    of the abstract methods such as :meth:`~nemos.observation_models.Observations.log_likelihood`,
+    :meth`~nemos.observation_models.Observations.sample_generator`, and
```
what's the tilde notation mean?

BalzaniEdoardo replied:
Tilde changes the repr, you don't have the full path but only `sample_generator`, at least that's how it woks with the classes. I have the same syntax in a `See Also` with GammaObservations

billbrod commented:
I also am fine with having multiple entries under raises, don't have a big preference for the numpydoc way of doing it

sjvenditto commented on docs/how_to_guide/plot_02_glm_demo.md:
````diff
@@ -104,53 +127,61 @@
-# %%
-# Hyperparameters can be set at any moment via the `set_params` method.
 
+```{code-cell} ipython3
````
This code cell additionally prints a UserWarning for regularizer strength. Should it be suppressed?

sjvenditto commented on docs/how_to_guide/plot_05_sklearn_pipeline_cv_demo.md:
````diff
@@ -0,0 +1,560 @@
+Let's start by creating some toy data.
+
+
+```{code-cell} ipython3
````
This prints a UserWarning for using plotting functions in `_documentation_utils`. Suppress?

sjvenditto commented on docs/how_to_guide/plot_06_glm_pytree.md:
````diff
@@ -9,133 +26,161 @@
+with `pip install dandi pynapple`
+:::
 
+```{code-cell} ipython3
````
prints UserWarning

sjvenditto commented on docs/how_to_guide/plot_06_glm_pytree.md:
````diff
@@ -144,13 +189,14 @@
+looks tuned for both.
 
+
+```{code-cell} ipython3
````
RuntimeWarning

sjvenditto commented on docs/how_to_guide/plot_06_glm_pytree.md:
````diff
@@ -216,15 +265,15 @@
+firing rates with that estimated from the counts.
 
 
+```{code-cell} ipython3
````
UserWarning

sjvenditto commented on docs/how_to_guide/plot_06_glm_pytree.md:
````diff
@@ -240,59 +289,66 @@
-# We could do all this with matrices as well, but we have to pay attention to
-# indices in a way that is annoying:
 
+```{code-cell} ipython3
````
UserWarning

sjvenditto commented on docs/how_to_guide/plot_02_glm_demo.md:
````diff
@@ -1,11 +1,29 @@
+---
+
+```{code-cell} ipython3
+%matplotlib inline
````
should we hide the matplotlib inline code blocks in the how-tos and tutorials?

```suggestion
:tags: [hide-cell]
%matplotlib inline
```

sjvenditto reviewed (changes requested):
Summary of suggested changes:
- warning admonition in installation.md needs to be reformatted
- hide code cell `%matplotlib inline` throughout tutorials and how_to_guide
- suppress warnings throughout tutorials and how_to_guide (I marked where I saw warnings in how_to_guide, I didn't do this for tutorials)

sjvenditto commented on docs/tutorials/plot_01_current_injection.md:
```diff
@@ -0,0 +1,739 @@
+pulse multiplied by a sinusoid of a fixed frequency, with some random noise
+riding on top.
+
+![Allen Brain Atlas view of the data we will analyze.](../../assets/allen_data.png)
```
this image isn't showing up

sjvenditto commented on docs/tutorials/plot_03_grid_cells.md:
````diff
@@ -1,70 +1,92 @@
-#
-# Here we load the data from OSF. The data is a NWB file.
 
+```{code-cell} ipython3
````
tqdm also renders poorly here

sjvenditto commented on docs/tutorials/plot_04_v1_cells.md:
````diff
@@ -0,0 +1,334 @@
+
+
+
+```{code-cell} ipython3
````
tqdm renders poorly here. readthedocs build also shows a KeyboardInterrupt error

sjvenditto commented on docs/tutorials/plot_05_place_cells.md:
````diff
@@ -0,0 +1,494 @@
+Here we load the data from OSF. The data is a NWB file.
+
+
+```{code-cell} ipython3
````
tqdm and KeyboardInterrupt error on readthedocs

sjvenditto commented on docs/tutorials/plot_06_calcium_imaging.md:
````diff
@@ -0,0 +1,345 @@
+Here we load the data from OSF. The data is a NWB file.
+
+
+```{code-cell} ipython3
````
tqdm rendered poorly on readthedocs here

sjvenditto reviewed (changes requested):
Found a few more issues when looking through the readthedocs build, it looks like sometimes fetching data fails in a couple of the tutorials

BalzaniEdoardo commented:
Wow this touched everything. I've got notes here and sprinkled throughout. I occasionally noticed things that weren't rendered correctly, but it's hard to do a systematic review of it. Maybe we need to each take a main .py file and look through their API docs during our meeting this week? It's hard to look at too many of them.

I tried to look through the tutorials etc and make sure they all looked correct. But it was fairly cursory.

- I believe there's a way to use sphinx's [cross-referencing ability](https://www.sphinx-doc.org/en/master/usage/domains/python.html#python-xref-roles) so we can avoid the manual writing of things like [`fit`](nemos.glm.GLM.fit) (though I haven't figured it out for plenoptic yet)

It may, but I think this is already a good enough solution

- Not sure how next/previous buttons are determined, but often the name of them is not helpful. e.g., on [Getting Help](https://nemos--264.org.readthedocs.build/en/264/getting_help.html), "Next" says "Contents" (because that's the first header), instead of "Developer Notes"
- 
I'll move between hd1 levels


- We may want a picture per card for the "Background" and "how to guide" pages. otherwise they look kinda blank

True, how should I tag an image? maybe for later

- On `api_reference.rst`, each block has a `.. current module::` and `.. autosummary::` block. what are they doing? (I had assumed that toctree would allow things to be automatically generated, but they're not, since the lines with `GLM` and `PopulationGLM` are still necessary)

Autosummary generates the table, current module tells autosummary to look for classes (or whatever listed) in a specific module. Without it you get
```
WARNING: [autosummary] failed to import GLM.
Possible hints:
* KeyError: 'GLM'
* ModuleNotFoundError: No module named 'GLM'
* ValueError: not enough values to unpack (expected 2, got 1)
```
and no table in the docs.

- `+++` appear around several of the notebooks. I don't think they do anything, should be removed

This were added by jupytext, I removed them all from the myst-nb files

- Two notes about the header:
  - If you zoom in enough that the text can't fit on a single line, it doesn't look good: 
![image](https://github.com/user-attachments/assets/0e8aace1-4cab-41ab-b32c-57e667f857fc)
  - must be some way to configure that to look better
  - is it possible to get rid of the More> dropdown? can we make it work so they all fit on a single line? changing `header_links_before_dropdown` does this, but interacts with the above point
  That's possible, it's a config
- tqdm output rendering is bad. I fixed this [here](https://workshops.plenoptic.org/workshops/CSHL-vision-course-2024/branch/main/introduction.html) but don't remember how. Maybe just using `from tqdm.auto import tqdm` in the underlying code? 
probably solved
- why are some methods (e.g., base_regressor.py > `_get_optimal_solver_params_config`) switched to private?
keep it private,  user should not interact
- References in API doc don't render well in dark mode: 
![image](https://github.com/user-attachments/assets/da0737a2-509d-4f91-99d1-89f52f35b123). 
  - probably some css would fix that? what about the line that's there?
- not all of the multiple exceptions have been fixed (e.g., `AdditiveBasis.evaluate_on_grid`)
- `basis.split_by_feature` also renders incorrectly. in general, it looks like there's an issue wherever there was a list.

BalzaniEdoardo commented:
on the pydata theme configurations: I doubt that they have a good solution for the header links since the home page of the pydata website has the same rendering issue
<img width="970" alt="Screenshot 2024-11-21 at 9 07 53 AM" src="https://github.com/user-attachments/assets/f8a16881-f8f2-40ff-be93-4b4a3e7d537a">

BalzaniEdoardo commented:
I opened a discussion on [PyData](https://github.com/pydata/pydata-sphinx-theme/discussions/2050).
hopefully there is a solution  already, or they open an issue about it.

BalzaniEdoardo commented on docs/background/plot_00_conceptual_intro.md:
```diff
@@ -0,0 +1,230 @@
+:::
+
+Specifically, our firing rate is:
+$$ \lambda (t) = \exp (L(x(t)) = \exp (w x(t) + c) \tag{2}$$
```
I kept it centered but now it renders.

BalzaniEdoardo commented on docs/conf.py:
```diff
@@ -0,0 +1,155 @@
+
+project = 'nemos'
+copyright = '2024'
+author = 'E Balzani'
```
Sure, I don't mind that we should have a discussion about this type of things. I think @gviejo  is more of the opinion that one would benefit more if credited individually on something like the landing page. I see the point as well, because people will read your name and associate it to the project, while "nemos authors" is more fare but needs an extra step to figure out who these people are.

BalzaniEdoardo commented on docs/conf.py:
```diff
@@ -0,0 +1,155 @@
+
+
+# Path for static files (custom stylesheets or JavaScript)
+html_static_path = ['assets/stylesheets', "assets"]
```
probably not

BalzaniEdoardo commented on docs/how_to_guide/plot_06_glm_pytree.md:
````diff
@@ -9,133 +26,161 @@
-# %%
-# We can map lambda functions:
+
+```{code-cell} ipython3
````
added to an issue

BalzaniEdoardo commented on docs/conf.py:
```diff
@@ -0,0 +1,155 @@
+from pathlib import Path
+
+sys.path.insert(0, str(Path('..', 'src').resolve()))
+sys.path.insert(0, os.path.abspath('sphinxext'))
```
unsure, this is what pynapple is doing; I'll leave it for noe
